### PR TITLE
[1410] Fix S3DynamoDBLogStore concurrent writer bug

### PR DESCRIPTION
**Pull request by scottsand-db**
## Description

See the full bug explanation at #1410, also [this comment in the code explains the problem scenario further](https://github.com/delta-io/delta/pull/1416/commits/5ef0035fd210b3065e64b5f8a9adff41854b98be#diff-50c09f154175b53894ded5fe6a722f12129ab3742afc7612f56fd6eebfe3487dR52-R80).

The existing code had the following bugs and we add the following fixes
- the copy `T(N)` into `N.json` code path in `BaseExternalLogStore` was using `overwrite=true`. This can overwrite existing files and cause data loss. This PR sets `overwrite=false`.
- when "acknowledging" the commit and writing to the external store entry `E(N, complete=true, commitTime=now)`, this let any underlying TTL infra clean up that commit **right away**, because our docs specified that the TTL key was the `commitTime` attribute. Instead, we need to add some delay (this PR uses 1 hour) after which the entry can be TTLd and cleaned up. This gives enough time that any future `fileSystem.exists(path)` call will return true and we don't run into any mutual exclusion issues. This PR also renamed this field to `expireTime`.
- in general, we were not short-circuiting early if `overwrite=false` and path already exists. This PR now does that.

## How was this patch tested?

Added new unit tests. Also, updated integration tests.

Ran integration tests with
```
# 1. Created DDB table
aws dynamodb create-table \
    --region us-west-2 \
    --table-name scott_test_2022_10_07_000 \
    --attribute-definitions AttributeName=tablePath,AttributeType=S \
                            AttributeName=fileName,AttributeType=S \
    --key-schema AttributeName=tablePath,KeyType=HASH \
                AttributeName=fileName,KeyType=RANGE \
    --provisioned-throughput ReadCapacityUnits=5,WriteCapacityUnits=5

# 2. Enabled TTL on that table
aws dynamodb update-time-to-live \
  --region us-west-2 \
  --table-name scott_test_2022_10_07_000 \
  --time-to-live-specification "Enabled=true, AttributeName=expireTime"

# 3. Configured test params
export DELTA_CONCURRENT_WRITERS=20
export DELTA_NUM_ROWS=50
export DELTA_DYNAMO_TABLE_NAME=scott_test_2022_10_07_000

# 4. Ran integration test
./run-integration-tests.py --use-local \
    --run-storage-s3-dynamodb-integration-tests \
    --dbb-packages org.apache.hadoop:hadoop-aws:3.3.1 \
    --dbb-conf io.delta.storage.credentials.provider=com.amazonaws.auth.profile.ProfileCredentialsProvider \
               spark.hadoop.fs.s3a.aws.credentials.provider=com.amazonaws.auth.profile.ProfileCredentialsProvider
```

## Does this PR introduce _any_ user-facing changes?

Users will need to modify any existing TTL settings to use the new field of `expireTime` instead of the old field `commitTime`.

### Enabling TTL for the first time
If TTL is not enabled but you would like to enabled it, perform
```
aws dynamodb update-time-to-live \
  --region <region> \
  --table-name <table-name> \
  --time-to-live-specification "Enabled=true, AttributeName=expireTime"
```
You may wish to clean up old, existing entries, too. It is safe to clean up any entry **older than an hour** with attribute `commitTime` set. Any entry with this attribute set will already be completed and written to the file system. Since it is older than an hour, the new fixes in this PR will mean we no longer need DynamoDB to provide the mutual exclusion during the write.

### Changing TTL from `commitTime` attribute to the new `expireTime` attribute
[As the DDB docs say](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/time-to-live-ttl-before-you-start.html)
> You cannot reconfigure TTL to look for a different attribute. You must disable TTL, and then reenable TTL with the new attribute going forward.

so first disable the current TTL attribute
```
aws dynamodb update-time-to-live \
  --region <region> \
  --table-name <table-name> \
  --time-to-live-specification "Enabled=false, AttributeName=commitTime"
```
and then enable the new TTL attribute
```
aws dynamodb update-time-to-live \
  --region <region> \
  --table-name <table-name> \
  --time-to-live-specification "Enabled=true, AttributeName=expireTime"
```

After this, you can perform the same cleanup as described above, for all entries older than an hour with a `commitTime` attribute set.

### CODE REVIEW

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:234]: We don't need to check this. The `debug` method we call below will check it for you.

Here is an example when we should check the flag:
```
if (LOG.isDebugEnabled()) {
   String path = // some heavy work
   LOG.debug("writeActions to: {}", path);
}
```

Same for other log changes.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:370]: Do we have any place converting `org.apache.hadoop.fs.FileAlreadyExistsException` thrown from here to `DeltaErrors.fileAlreadyExists`?

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:141]: This check is not needed. Right? Are you just trying to avoid touching DynamoDBLogStore if the file already exists?

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:369]: nit: `input_stream` -> `inputStream`, `output_stream` -> `outputStream`

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:82]: Can we use 1 day for safety? It's better to add a flag for this value as well.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/ExternalCommitEntry.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/ExternalCommitEntry.java:59]: We can keep `commitTime`. It will be helpful to help us debug issues.

**Review by zsxwing — COMMENTED** [review]
A side question: when do we delete the temp file after copy?

**Review by ormahler — CHANGES_REQUESTED** [review]

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:141]: > This validation better be as close as possible to the part where it's added to the external log-store.

Why? We have changed our TTL such that it is an hour (but @zsxwing wants it changed to a day) after the write. We assume that our file system cannot provide mutual exclusion on writes .... but it can detect after several minutes (and therefore, also for an hour or a day) if a file already exists in the FileSystem.

Comment (ormahler) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:141]: The new ttl does solve it.
But for the sake of users who upgraded delta but didn’t notice they need to update the ttl definition, better minimize the risky time

**Comment by zsxwing** [comment]
Deleting based on the temp file creation time is not safe. If the user stops a write query before copying the temp file to the dest location, and does nothing in 30 days, the table will be broken.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:141]: That's a good point! Yep dropping `commitTime` is safer. We also need to write a migration guide for the next release to highlight the recommended TTL change.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:234]: sounds good.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:370]: Hi @ormahler I'm aware of this.
> Take into account 

Any suggestion?

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:141]: Gotcha. I just wanted to confirm this is about optimization rather than correctness.

Comment (zsxwing) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:141]: Gotcha. I just wanted to confirm this is about optimization rather than correctness.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:370]: No. `delta-core` depends on `delta-storage`. `storage-s3-dynamodb` also depends on `delta-storage`, it can't access code from `delta-core`.

But we should convert this error here it to `java.nio.file.FileAlreadyExistsException`

Comment (ormahler) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java]: Actually with the current implementation it’s not expected to occur. 
Just left it as another reason to keep line 141 and not to rely on this one.

**Review by ormahler — APPROVED** [review]

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:370]: we need to put `fs.create`

File: storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java
Comment (scottsand-db) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java:119]: use `hadoopConf`

**Review by ormahler — APPROVED** [review]

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:375]: I realized that whether changing `overwrite` to `false` doesn't matter for correctness because `copyFile` is called from two methods `fixDeltaLog` and `write`, and both of them will ignore any `Throwable`.

There are two cases we may hit `FileAlreadyExistsException` here:

1. `fixDeltaLog` is called at the same time when a write is completing the write. `FileAlreadyExistsException` will be ignored by `fixDeltaLog` properly. The writer side should also ignore it. I guess this is why it's `overwrite=true` previously.
2. `expirationDelaySeconds` is set to a small value such as 0. In this case, we need to notify the writer that this file does exist.

It seems we don't have any information in the writer side to distinguish these two cases. However, the default `expirationDelaySeconds` is 1 day which should be long enough to prevent Case 2.

Changing it to `false` may be still helpful for performance because we don't need to upload the file if the file already exists.

In a second thought, maybe we should ignore `FileAlreadyExistsException` here? Otherwise, `fixDeltaLog` and `write` may log a lot of `FileAlreadyExistsException`s, which is confusing.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:147]: Now I think this check is necessary. Without it, the only protection is just coming from DymamoDB. If the entry gets removed from DymamoDB for whatever reason, anyone can overwrite an existing file.

Comment (ormahler) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:147]: I still think this part should be as close as possible to the dynamoDB check, in order to minimize the risky time in which the entry can be removed from ddb.

Comment (zsxwing) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:147]: That's a good point. +1

Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:147]: > I still think this part belongs as close as possible to the dynamoDB check, in order to minimize the risky time in which the entry can be removed from ddb.

But if we already set the TTL delay was 1 day, I don't see why this code optimization is necessary?

Plus, with what you suggested (moving it closer), then we will still be creating the temp file and an entry in DDB, which is unecessary.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:70]: Yes, exactly.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:375]: > In a second thought, maybe we should ignore FileAlreadyExistsException here? Otherwise, fixDeltaLog and write may log a lot of FileAlreadyExistsExceptions, which is confusing.

both of those APIs will first check if the destination file exists first before trying to copy. so do you really think they may log a lot of `FileAlreadyExistsExceptions`?

**Comment by scottsand-db** [comment]
> Deleting based on the temp file creation time is not safe. If the user stops a write query before copying the temp file to the dest location, and does nothing in 30 days, the table will be broken.

If we do the deletion ourselves, that will cost users money (S3 API calls). s3 object expiration is free.

should we add a flag that users can set where we DO delete the temp file? After the write has been acknowledged.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:382]: That's fair

**Review by zsxwing — APPROVED** [review]
LGTM

**Comment by zsxwing** [comment]
> should we add a flag that users can set where we DO delete the temp file? After the write has been acknowledged.

Let's solve this separately. We don't need to block this PR for this rare issue.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (zsxwing) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:151]: Let's leave it as it is. Commit conflicts would be more common than incorrect short TTL delay config.

Comment (ormahler) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:151]: I agree that it’s unnecessary given 1 day ttl, but can we be sure it’s kept?

anyway my suggestion is to place it before the insert to ddb so no extra entry will be entered.
I don’t think creating a temp file is a big deal but you can also place it before it,
just after that part:
`// Step 1: Ensure that N-1.json exists`

File: storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java
Comment (ormahler) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java:123]: Worth to define a minimal expiration time larger thank 0 and validate it here.

Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java:123]: What should that minimal expiration time be?

Comment (zsxwing) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java:123]: We don't plan to document this flag. It's mostly just a best practice to avoid a hard-code value. @scottsand-db we can add a comment for the config key to warn the user.

File: storage-s3-dynamodb/integration_tests/dynamodb_logstore.py
Comment (zsxwing) [inline storage-s3-dynamodb/integration_tests/dynamodb_logstore.py:128]: nit: remove this as we don't want users to use this.

File: storage-s3-dynamodb/integration_tests/dynamodb_logstore.py
Comment (zsxwing) [inline storage-s3-dynamodb/integration_tests/dynamodb_logstore.py:108]: nit: remove this as we don't want users to use this.

File: storage-s3-dynamodb/integration_tests/dynamodb_logstore.py
Comment (zsxwing) [inline storage-s3-dynamodb/integration_tests/dynamodb_logstore.py:81]: nit: remove this as we don't want users to use this.

File: storage-s3-dynamodb/integration_tests/dynamodb_logstore.py
Comment (zsxwing) [inline storage-s3-dynamodb/integration_tests/dynamodb_logstore.py:56]: nit: remove this as we don't want users to use this.

File: storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java
Comment (zsxwing) [inline storage-s3-dynamodb/src/main/java/io/delta/storage/S3DynamoDBLogStore.java:88]: nit: add a comment to warn the user that they should not set this.

**Review by zsxwing — APPROVED** [review]

**Comment by scottsand-db** [comment]
@ormahler thanks so much for the reviews! Can I include you as a coauthor of this PR? Considering that I used your original PR as a starting point.

**Comment by ormahler** [comment]
> @ormahler thanks so much for the reviews! Can I include you as a coauthor of this PR? Considering that I used your original PR as a starting point.

With pleasure!
That would be great, thanks a lot.

**Comment by scottsand-db** [comment]
@ormahler - great, can you give me your github committer email address? I'll need to push another commit taking you as a coauthor.

**Comment by ormahler** [comment]
@scottsand-db 
ormahler31@gmail.com

File: storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java
Comment (Arnovsky) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:71]: @scottsand-db Trying to verify my understanding of this comment, does this mean TTL can be set to 0 if using S3 as storage due to the fact that it's strongly consistent?

Comment (scottsand-db) [reply storage-s3-dynamodb/src/main/java/io/delta/storage/BaseExternalLogStore.java:71]: I wouldn't encourage that.

S3 doesn't provide mutual exclusion on writes. We need to ensure that two concurrent writers have some sufficient delta time between the two writes to avoid any mutual exclusion issues. If you set the TTL to 0 ... then you are just getting closer and closer to a mutual exclusion problem, with what benefit?

**Pull request closed without merging**
